File: src/main/ui/src/models/core/PDTypeMeta.ts
import type {PDMeta} from "@/models/core/PDMeta";
import type {PDAttributeMeta} from "@/models/core/PDAttributeMeta";
import type {PDOperationMeta} from "@/models/core/PDOperationMeta";
import type {PDAssociationMeta} from "@/models/core/PDAssociationMeta";
import type {MultiLanguageLabel} from "@/models/core/MultiLanguageLabel";
import type {ObjectStringifier} from "@/models/core/ObjectStringifier";

export interface PDTypeMeta extends PDMeta {
    typeId: Number;
    labelPlural: MultiLanguageLabel;
    icon?: string;
    objectStringifier: ObjectStringifier;
    subsystem?: string;
    base?: PDTypeMeta;
    attributes: PDAttributeMeta[];
    associations?: PDAssociationMeta[];
    operations?: PDOperationMeta[];
}

File: src/main/java/de/fhswf/genericapplication/models/core/PDTypeMeta.java
package de.fhswf.genericapplication.models.core;

import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.databind.annotation.JsonSerialize;
import de.fhswf.genericapplication.models.BaseEntity;
import de.fhswf.genericapplication.serializers.MultiLanguageLabelSerializer;

import javax.xml.bind.annotation.*;
import java.util.ArrayList;
import java.util.List;

@XmlRootElement(namespace = "core", name = "type")
@XmlAccessorType(XmlAccessType.FIELD)
public class PDTypeMeta extends PDMeta {
    @XmlAttribute
    @JsonProperty
    private Long typeId;

    @XmlAttribute
    @JsonProperty
    @JsonSerialize(using = MultiLanguageLabelSerializer.class)
    private String labelPlural;

    @XmlAttribute
    @JsonProperty
    private String icon;

    @XmlAttribute
    @JsonProperty
    private String objectStringifier;

    @XmlAttribute
    @JsonProperty
    private String subsystem;

    @XmlAttribute
    @JsonProperty
    private String base = BaseEntity.class.getSimpleName();

    @XmlAttribute
    @JsonProperty
    private boolean singleton = false;

    @XmlAttribute
    @JsonProperty
    private boolean masterData = false;

    @XmlElementWrapper(namespace = "core", name = "attributes")
    @XmlElement(namespace = "core", name = "attribute")
    @JsonProperty
    private List<PDAttributeMeta> attributes = new ArrayList<>();

    @XmlElementWrapper(namespace = "core", name = "associations")
    @XmlElement(namespace = "core", name = "association")
    @JsonProperty
    private List<PDAssociationMeta> associations = new ArrayList<>();

    @XmlElementWrapper(namespace = "core", name = "operations")
    @XmlElement(namespace = "core", name = "operation")
    @JsonProperty
    private List<PDOperationMeta> operations = new ArrayList<>();
}
